File: apps/desktop/src/renderer/src/lib/api/endpoints/org.ts
/** org chart endpoints — the chain of command between agents. */
import { requestJson } from "../http";
import type { OrgActionResult, OrgChartResponse, OrgLinkItem } from "../types";

/** `GET /org/chart` — agents + saved reporting lines. */
export function getOrgChart(): Promise<OrgChartResponse> {
  return requestJson<OrgChartResponse>("/org/chart", { timeoutMs: 40_000 });
}

/** `PUT /org/chart` — validate, save, and APPLY the chart (permissions + team
 *  notes + gateway reload). May restart the gateway, hence the long timeout. */
export function saveOrgChart(links: OrgLinkItem[]): Promise<OrgActionResult> {
  return requestJson<OrgActionResult>("/org/chart", {
    method: "PUT",
    body: { links },
    timeoutMs: 120_000,
  });
}

File: services/agent-engine/app/api/org.py
"""Org chart endpoints — the chain of command between OpenClaw agents."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from ..schemas.org import OrgActionResult, OrgChartResponse, OrgSaveRequest
from ..services.org_service import OrgService
from .deps import get_org_service

router = APIRouter(prefix="/org", tags=["org"])


@router.get("/chart", response_model=OrgChartResponse, summary="The org chart")
def get_org_chart(service: OrgService = Depends(get_org_service)) -> dict:
    """All OpenClaw agents (with identity) + the saved reporting lines."""
    return service.get_chart()


@router.put("/chart", response_model=OrgActionResult, summary="Save + apply the org chart")
def save_org_chart(
    payload: OrgSaveRequest,
    service: OrgService = Depends(get_org_service),
) -> dict:
    """Validate the tree (one manager each, no loops), persist it, and apply it to
    OpenClaw: per-manager delegation allowlists, team notes in AGENTS.md, and a
    safe spawn depth. Managers can then command their teams — including from Discord."""
    return service.save_chart([l.model_dump() for l in payload.links])
